catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/ws/JsonRpc.java

```diff
@@ -0,0 +1,178 @@
+
+// An implementation of http://www.jsonrpc.org/specification over websockets
+public class JsonRpc implements Socket {
+
+  public static final String VERSION = "2.0";
+
+  public static final int PARSE_ERROR = -32700;
+  public static final int INVALID_REQUEST = -32600;
+  public static final int METHOD_NOT_FOUND = -32601;
+  public static final int INVALID_PARAMS = 32602;
+  public static final int INTERNAL_ERROR = -32603;
+
+  private ObjectMapper mapper =
+      JsonFactory.create(
+          new JsonParserFactory().usePropertyOnly(),
+          new JsonSerializerFactory()
+              .includeEmpty()
+              .includeNulls()
+              .includeDefaultValues()
+              .setJsonFormatForDates(false));
+
+  private final Map<String, Function> methods;
+
+  public JsonRpc(Map<String, Function> methods) {
+    this.methods = methods;
+  }
+
+  private static Map<String, Object> parseError(Object message) {
+    return error(PARSE_ERROR, "Parse error", message);
+  }
+
+  private static Map<String, Object> error(int code, String message) {
+    return error(code, message, null);
+  }
+
+  private static Map<String, Object> error(int code, String message, Object data) {
+    Map<String, Object> response = new HashMap<>();
+    response.put("code", code);
+    response.put("message", message);
+    if (data != null) {
+      response.put("data", data);
+    }
+    return response;
+  }
+
+  private static Map<String, Object> success(Object id, Object result) {
+    Map<String, Object> response = new HashMap<>();
+    response.put("jsonrpc", VERSION);
+    response.put("id", id);
+    response.put("result", result);
+    return response;
+  }
+
+  private static Map<String, Object> fail(Object id, Object error) {
+    Map<String, Object> response = new HashMap<>();
+    response.put("jsonrpc", VERSION);
+    response.put("id", id);
+    response.put("error", error);
+    return response;
+  }
+
+  private Map<String, Object> invalid() {
+    return invalid(null);
+  }
+
+  private Map<String, Object> invalid(Object data) {
+    return error(INVALID_REQUEST, "Invalid Request", data);
+  }
+
+  private Map<String, Object> invalidParams(Object data) {
+    return error(INVALID_PARAMS, "Invalid params", data);
+  }
+
+  @Override
+  public void onOpen(Session session) {}
+
+  @Override
+  public void onClose(Session session, int statusCode, String reason) {}
+
+  @Override
+  public void onError(Session session, Throwable ex) {}
+
+  private Object exec(Object message) {
+
+    if (message instanceof List) {
+      return ((List) message).parallelStream().map(this::exec).collect(Collectors.toList());
+    }
+
+    if (!(message instanceof Map)) {
+      return invalid();
+    }
+
+    Map msg = (Map) message;
+
+    if (!msg.containsKey("id")) {
+      return fail(null, invalid("Required key `id` missing"));
+    }
+
+    Object id = msg.get("id");
+
+    if (!(id instanceof String || id instanceof Number || id == null)) {
```

Comment:
❓ Should this be `!(id instanceof String || id instanceof Number) || id == null`?

Reply:
I think it's okay because we want to allow null values for id. 

![image](https://user-images.githubusercontent.com/1986211/36576078-a7673702-180b-11e8-9078-9af04eca837a.png)